### PR TITLE
Make RESTEasy Reactive handlers calls monomorphic

### PR DESCRIPTION
**Pull request by geoand**
This looks ugly, but (hopefully) results in better performance

### CODE REVIEW

**Comment by geoand** [comment]
@johnaohara would someone from your team be able to test this and see if it does indeed result in better performance in a proper lab scenario?

I did some tests of my own and does seem to lead to minor performance improvements, but without a real performance test, it's hard to know for sure.

**Comment by Plawn** [comment]
Hello, I don't know, if it's the right place to ask, this kind of question, but while looking at this PR, I tought about how mapstruct builds source code directly to avoid iterations and tailors a method to specifically do one thing.

Would it be possible to do something like this and produce "hardcoded" handlers at build time when enough parameters are known, to get even more performance ?

**Comment by geoand** [comment]
> Would it be possible to do something like this and produce "hardcoded" handlers at build time when enough parameters are known, to get even more performance ?

Do you an example in mind?

**Comment by johnaohara** [comment]
@geoand We did something similar to this in Hibernate to force a direct method invocation, but after speaking to andrew hayley there were other options available to avoid the virtual method invoke, i'll take a look

**Comment by Plawn** [comment]
> > Would it be possible to do something like this and produce "hardcoded" handlers at build time when enough parameters are known, to get even more performance ?
> 
> Do you an example in mind?

After thinking more about it, It would be like a a loop unrolling but, by knowing the types of the handlers, we could force the types of the handlers at compile and avoid instanceOf checks:
maybe it' s dumb :)

After reading even more code, I see, that it can't be implemtented due to the non final nature of the handlers array and I'm wondering if the restart method, really needs to erase the handlers array.

(sorry guys if I'm saying dumb things, I'm new to the codebase, but I really like it)

**Comment by geoand** [comment]
@Plawn not dumb at all! Glad to see you reading the code and figuring out things 😉

**Comment by johnaohara** [comment]
@geoand thinking about this some more, there are a number of considerations around this change;

1. You detected a positive performance impact for the tests that you ran, but am wondering what the performance improvement would be based on different length handler chains, different patterns of handler chains etc. i.e. Does the branching overhead start to become larger than the virtual method dispatch overhead under certain conditions
2. There is a maintenance overhead of this code, i.e. if new handlers are added, they will default to the "slow path" unless explicitly defined in the switch statement
3. There may be something we can do here with with lambdas to force a direct method invocation (something along the lines @Plawn is suggesting), rather than making the call through the interface method

I will create a simple benchmark with different impl options (and access patterns) and we can see what would work best

**Comment by geoand** [comment]
> I will create a simple benchmark with different impl options (and access patterns) and we can see what would work best

Looking forward to it :)

**Comment by johnaohara** [comment]
@Plawn  

> Maybe we can directly reuse an application written using Quarkus which represents the average Quarkus application ?

Are you talking about using a quarkus application to test the different options? 

When I mentioned,

> create a simple benchmark

I was referring to creating a JMH microbechmark that tests different access handler patterns/chain lengths to allow us to understand what would be the best choice in general, rather than a solution to a specific application.

An **average** application is hard to define

**Comment by Plawn** [comment]
@johnaohara 
> Are you talking about using a quarkus application to test the different options?

Yes, it would present a real world application. Maybe the cost of function call indirection / cost of virtual call can be measured in synthetic benchmark, but if its not measurable in real usage, is it worth it ?
I'm not tech lead on this code base, so I'm not the one making a choice, I'm just wondering if it's the kind of optimization that is really worth it.

For a framework I wrote in python, i wanted to make this kind of optimization.
I wrote a library that would produce source code, compile it and replace the function to adapt function prototype, in order to avoid using the **kwargs operator. 
I did this because I measured that my technic was between 30% and 20% faster.

It worked, and on that function call it was faster, but as a whole it was neglieable, if I remember correctly.

just a thought

**Comment by geoand** [comment]
Thanks for the excellent analysis @franz1981! 

The types I have added in this PR are basically the ones that will most likely be used during the servicing of a request (most will  actually always be used, some will likely be used).
The idea is that each handler is a processing step and depending on what needs to be done, various handlers are added to the processing pipeline - which is why some are always executed (essentially the first ones as you read the changes in the PR).

**Comment by geoand** [comment]
A very simple application that should be somewhat indicative of real world use, is https://github.com/quarkusio/quarkus-quickstarts/tree/main/rest-json-quickstart

**Comment by franz1981** [comment]
I'll take a deeper look at https://github.com/quarkusio/quarkus-quickstarts/tree/main/rest-json-quickstart, in the meantime worth checking what's the bytecode size of the method that contains the instanceof check chains: I suspect is too big to be inlined...

image

Linux byte code size for frequent calls is 325 bytes and it would be nice of it get inlined as well, if possible.

The other thing I'm not fully convinced is the behaviour of the JIT with so many checks that could be re-arranged (with deoptimizations) depending the actual conditional branches taken at runtime. I'll paste here any finding

**Comment by geoand** [comment]
> in the meantime worth checking what's the bytecode size of the method that contains the instanceof check chains: I suspect is too big to be inlined

I didn't think about that...
So I wonder if I should only keep some of the checks.

**Comment by franz1981** [comment]
@geoand @johnaohara I have to check if this https://github.com/forax/exotic#typeswitch---javadoc can work as expected without using a cascade of instance of and turning the cascade of instanceof into a switch (as suggested elsewhere)

**Comment by geoand** [comment]
Cool, looking forward to it

**Comment by geoand** [comment]
@franz1981 I looked into the `TypeSwirth` thing from `exotic` and I really liked what I saw - very cool stuff, I learned new things :)

The problem however is that it won't work in native mode... We could certainly have a different code path for native mode, but unless `TypeSwitch` provides a noticeable throughput improvement, I would rather leave it out.

**Comment by franz1981** [comment]
> @franz1981 I looked into the TypeSwirth thing from exotic and I really liked what I saw - very cool stuff, I learned new things :)

Me too, but I've sadly discovered that it doesn't work very well in term of performance
I've written a micro bench handling exactly the same number of impls of this PR and these are my results:
```

Benchmark                                                       (samples)  (top)  (types)  Mode  Cnt    Score   Error  Units
ITableMegamorphicTests.do_Dynamic_Interface_Ref                     10000  false       10  avgt   10  289.776 ± 2.049  us/op
ITableMegamorphicTests.do_Dynamic_Interface_Ref                     10000   true       10  avgt   10  291.460 ± 2.946  us/op
ITableMegamorphicTests.do_Interface_ID_switch                       10000  false       10  avgt   10  271.008 ± 2.292  us/op
ITableMegamorphicTests.do_Interface_ID_switch                       10000   true       10  avgt   10  259.129 ± 3.213  us/op
ITableMegamorphicTests.do_Peel_Interface_Interface                  10000  false       10  avgt   10  240.129 ± 1.469  us/op
ITableMegamorphicTests.do_Peel_Interface_Interface                  10000   true       10  avgt   10  229.023 ± 3.762  us/op
ITableMegamorphicTests.do_TypeSwitch_ClassValue_Interface           10000  false       10  avgt   10  458.165 ± 0.812  us/op
ITableMegamorphicTests.do_TypeSwitch_ClassValue_Interface           10000   true       10  avgt   10  446.159 ± 4.370  us/op
ITableMegamorphicTests.do_TypeSwitch_Interface                      10000  false       10  avgt   10  491.718 ± 1.006  us/op
ITableMegamorphicTests.do_TypeSwitch_Interface                      10000   true       10  avgt   10  376.023 ± 2.342  us/op
```
Source code on https://github.com/franz1981/java-puzzles/tree/main/src/main/java/red/hat/puzzles/polymorphism

TLDR

The results refer to uniform distributed 10 types vs 15 possible, where they can happen from the top or from the bottom (to stress the `instanceof` if chain in the two opposite directions).
`do_Peel_Interface_Interface `  (that's the strategy you've proposed) is a clear winner over `.do_Dynamic_Interface_Ref` that's the purely virtual itable dispatch (on 'main`).

Let me know if you need other details on the bench itself; I've discovered a couple of interesting things too ie
- `ClassValue` isn't that fast with interfaces (they end up hitting a slow path there): that's what Remi Forax has implemented
- JDK 11 bumped the `MinJumpTableSize` from 10 to 18 (see how is used on [JDK 11 source ode](https://github.com/openjdk/jdk11/blob/37115c8ea4aff13a8148ee2b8832b20888a5d880/src/hotspot/share/opto/parse2.cpp#L841)): TLDR if there are *more* then 18 `case`, a `switchtable` construct is compiled (in assembly) as a pointer table (not inlined in the caller assembly, but separate), while with less `case` it's compiled as a binary search over a chain of `if` condition, inlined in the caller: this can make some difference while deciding if use more or less cases in the switch (see more on this [old but very detailed explanation](https://exchangetuts.com/why-does-java-switch-on-contiguous-ints-appear-to-run-faster-with-added-cases-1639470184483103)) 

Today I plan to run a bench with some real usage, just to verify anything bad isn't happening, but right now it seems a very beneficial change even in the current form (and in the microbenchmark, the bytecode size of the method seems enough to getting it inlined too)

**Comment by franz1981** [comment]
> I would rather leave it out

Fully agree and according to the results is not even that fast, so...2 pigeons with 1 stone (poor pigeons!)

**Comment by geoand** [comment]
Wonderful information @franz1981, thanks a ton!

**Comment by geoand** [comment]
I marked this as ready for review

**Review by franz1981 — APPROVED** [review]

**Comment by geoand** [comment]
CC @Sanne

**Review by Sanne — APPROVED** [review]
